### PR TITLE
LoRA レイヤー別学習率の実装、state_dict読み込みの際のdevice指定削除、typo修正

### PR DESCRIPTION
**Pull request by u-haru**
# LoRA レイヤー別学習率の実装
`network_args`に`down_lr_weight`, `mid_lr_weight`, `up_lr_weight`の3つのオプションを用意します。指定された値と学習率の積を学習率とすることで、レイヤー毎に異なる学習率を適用します。また、weightが0のレイヤーはapplyしないように変更しています。現在実装してる引数は以下のようになります。
・down_lr_weight=0,0,0,0,0,1,1,1,1,1 (conv2d_3x3使用時)
・down_lr_weight=0,0.5,1 (conv2d_3x3未使用時)
・mid_lr_weight=0.5
・up_lr_weight=linear (0から1へと変化する。他にはsine, cosine, reverse_linear, zeros等)
・stratified_zero_threshold=0.1 (0.1以下のweightは0として扱う)

レイヤー構造は私自身があまり理解できていなかった為、以下のように階層化して適用しています(prefix等一部省略)。conv2d_3x3無効時はupとdownでそれぞれ3層、有効時は10層に分割しています。midは常に1層として扱います。

conv2d_3x3無効時
```
(1)down_blocks_0_attentions_{0,1}
(2)down_blocks_1_attentions_{0,1}
(3)down_blocks_2_attentions_{0,1}

mid_block_attentions_0
mid_block_resnets_{0,1}

(1)up_blocks_1_attentions_{0,1,2}
(2)up_blocks_2_attentions_{0,1,2}
(3)up_blocks_3_attentions_{0,1,2}
```

conv2d_3x3有効時
```
(1)down_blocks_0_resnets_{0,1}
(2)down_blocks_0_downsamplers
(3)down_blocks_0_attentions_{0,1}
(4)down_blocks_1_resnets_{0,1}
(5)down_blocks_1_downsamplers
(6)down_blocks_1_attentions_{0,1}
(7)down_blocks_2_resnets_{0,1}
(8)down_blocks_2_downsamplers
(9)down_blocks_2_attentions_{0,1}
(10)down_blocks_3_resnets_{0,1}

mid_block_attentions_0
mid_block_resnets_{0,1}

(1)up_blocks_0_resnets_{0,1,2}
(2)up_blocks_0_upsamplers
(3)up_blocks_1_attentions_{0,1,2}
(4)up_blocks_1_resnets_{0,1,2}
(5)up_blocks_1_upsamplers
(6)up_blocks_2_attentions_{0,1,2}
(7)up_blocks_2_resnets_{0,1,2}
(8)up_blocks_2_upsamplers
(9)up_blocks_3_attentions_{0,1,2}
(10)up_blocks_3_resnets_{0,1,2}
```

これによって、学習したい範囲(細かいディテールの部分、全体の色の塗り等)を限定して学習することが出来るようになります。  
例えば、`down_lr_weight=cosine mid_lr_weight=0 up_lr_weight=sine` とすればモデルの浅い層をメインに学習させることができます。
逆に、`down_lr_weight=sine mid_lr_weight=1 up_lr_weight=cosine` とすればモデルの深い層をメインに学習させられます。

以下に私が試した結果を示します。制服を着せたキャラクターをCM3D2で用意して、"test"というトークンに紐づくように800step程度学習させたLoRA(dim: 64, conv_dim:1, lr: 1e-4, text_encoder_lr: 5e-5)です。  
xyz_grid-0018-1204636062
(down_lr_weight:mid_lr_weight:up_lr_weight)という表記で書くと左から順に
デフォルト(1:1:1)、sine:1:cosine、1:0:1、cosine:0:sine、LoRA無しです。

誤差レベルかもしれませんが、デフォルトでは3Dゲームっぽい画像が生成されるようになる(ゲームのスクショを使ってるため)のに対し、深い層のlrを下げた方はデフォルトに比べて色の塗り方が維持されているように思えます。

# state_dict読み込みの際のdevice指定削除
更新した際に`safetensors.torch.load_file`が`Exception: device cuda is invalid`というエラーを返すようになっていたのでその修正です。

他の環境では試せていませんが、私の環境で学習ができなくなっていたことと「`state_dict`読み込み時は`device`を指定する必要が無いだろう」という判断から`state_dict`読み込みの際の`device`の指定を外しています。モデルに`state_dict`を読み込む際の`.to(device)`等はそのままなので、特に問題ないかと思われます(私の環境では問題なく動作しています)。

# typo修正
`繰り返し回数のないディレクトリを無視します`の表示がバグっていたので修正です。

### CODE REVIEW

**Comment by u-haru** [comment]
お忙しい中ありがとうございます。

影響範囲を考えるとPullRequest分けたほうが良かったですね…
以降気をつけます。

**Comment by u-haru** [comment]
ありがとうございます。層の指定については、リンク先の方を参考にして12層に分けるようにしました。
IN: 0 ~ 11
MID: 0
OUT: 0 ~ 11
(BASEはテキストエンコーダなので不要？)
- 指定方法に関しては以前の`network_args`の設定をそのまま引き継いでいます。
- レイヤー数の条件分岐が大変になるので、どのような学習設定でもフルモデル相当であると仮定して12層で処理するようにしています。
- LoRAではIN00が使用されていない(？)ようなのでこの層の設定は効果が無いですが、わかりやすさのためにそのままにしてます。

次にD-Adaptation optimizer使用時の動作に関してですが、レイヤー別学習率を使用しない際にはparam_dataにlrを付与しないように変更しました。恐らくこれで正常な動作になっているはずです。

最後にLoRAモジュールの扱いに関してです。
削除した際の利点として学習が早くなるという点があり、可能であれば不要なパラメータを削除したほうがいいのではないかと考えています。なるべくパラメータを残したまま不要な学習をしないようにしようと色々試してみたのですが、
- LoRANetWorkの`self.add_module()`が実行された時点で(`apply_to()`をしていなくても)学習が遅くなる
- `self.add_module()`せずに`apply_to()`だけ実行すると学習時にエラーが発生する
- optimizerの`trainable_params`に渡さなくても変化しない
- 使わないモジュールで`requires_grad_(False)`しても変化しない
といった感じでした。

学習スピードに関しては、私の環境では深い層を除いた際に1.6it/s -> 1.9it/s程度まで上昇しました。
効果が結構大きいので、とりあえず現時点ではパラメータを削除する実装のままになっています。

**Comment by u-haru** [comment]
あまり複雑なことをしたくなかったので、とりあえず条件分岐で分けました。
コードが冗長になってしまった気もしますが、層別学習率を使わない場合の動作は以前と同じになったと思います。

**Comment by u-haru** [comment]
マージ確認しました。お忙しい中対応していただき、ありがとうございました。
